File: backend/internal/renderer/surge.go
package renderer

import (
	"fmt"
	"subforge/internal/core"
)

type SurgeRenderer struct{}

func (r *SurgeRenderer) Name() string { return core.FormatSurge }

func (r *SurgeRenderer) Render(nodes []core.ProxyNode) (string, error) {
	out := "[Proxy]\n"
	for i, n := range nodes {
		out += renderSurgeProxy(n, i) + "\n"
	}
	return out, nil
}

func renderSurgeProxy(n core.ProxyNode, idx int) string {
	switch n.Type {
	case core.TypeVMess:
		line := fmt.Sprintf("%s = vmess, %s, %d, username=%s",
			n.Name, n.Server, n.Port, n.UUID)
		if n.TLS {
			line += ", tls=true"
			if n.SNI != "" {
				line += fmt.Sprintf(", sni=%s", n.SNI)
			}
		}
		return line

	case core.TypeTrojan:
		line := fmt.Sprintf("%s = trojan, %s, %d, password=%s",
			n.Name, n.Server, n.Port, n.Password)
		if n.SNI != "" {
			line += fmt.Sprintf(", sni=%s", n.SNI)
		}
		return line

	case core.TypeSS:
		return fmt.Sprintf("%s = ss, %s, %d, encrypt-method=%s, password=%s",
			n.Name, n.Server, n.Port, getStr(n.Extra, "cipher"), getStr(n.Extra, "password"))

	default:
		return fmt.Sprintf("# unsupported: %s = %s, %s, %d", n.Name, n.Type, n.Server, n.Port)
	}
}

File: backend-python/exporters/surge.py
import json
import logging
from typing import List, Dict, Any

logger = logging.getLogger(__name__)


def generate_surge_config(nodes: List[Dict[str, Any]]) -> str:
    """
    生成 Surge 配置格式

    格式: [Proxy]
    NodeName = protocol, server, port, password, opt1=value1, opt2=value2
    """
    lines = ["[Proxy]"]

    for node in nodes:
        node_type = node.get("type", node.get("node_type", "")).lower()
        name = node.get("name", "Unknown")
        server = node.get("server", "")
        port = node.get("port", 0)
        config = node.get("data", node.get("config_json", {}))

        if not server or not port:
            continue

        try:
            line = _generate_proxy_line(node_type, name, server, port, config)
            if line:
                lines.append(line)
        except Exception as e:
            logger.warning(f"Failed to generate Surge config for {name}: {e}")

    return "\n".join(lines)


def _generate_proxy_line(node_type: str, name: str, server: str, port: int, config: Dict[str, Any]) -> str:
    """生成单个代理配置行"""

    if node_type == "ss":
        method = config.get("cipher", config.get("method", "aes-256-gcm"))
        password = config.get("password", "")
        return f"{name} = ss, {server}, {port}, encrypt-method={method}, password={password}"

    elif node_type == "vmess":
        uuid = config.get("id", config.get("uuid", ""))
        alter_id = config.get("aid", 0)
        cipher = config.get("scy", "auto")
        tls = config.get("tls", False)
        line = f"{name} = vmess, {server}, {port}, username={uuid}, vmess-aes={cipher}"
        if tls:
            line += ", tls=true"
            if config.get("sni"):
                line += f", sni={config['sni']}"
        return line

    elif node_type == "vless":
        uuid = config.get("uuid", config.get("id", ""))
        tls = config.get("tls", False)
        flow = config.get("flow", "")
        line = f"{name} = vless, {server}, {port}, username={uuid}"
        if tls:
            line += ", tls=true"
            if config.get("sni"):
                line += f", sni={config['sni']}"
        if flow:
            line += f", vless-opts={flow}"
        return line

    elif node_type == "trojan":
        password = config.get("password", "")
        sni = config.get("sni", "")
        line = f"{name} = trojan, {server}, {port}, password={password}"
        if sni:
            line += f", sni={sni}"
        return line

    elif node_type == "hysteria2":
        password = config.get("password", "")
        line = f"{name} = hysteria2, {server}, {port}, password={password}"
        if config.get("sni"):
            line += f", sni={config['sni']}"
        return line

    return None
